doc/developer/design/20230329_monotonic_optimization.md

```diff
@@ -0,0 +1,177 @@
+-->
+
+We change monotonicity information from a boolean value to the following enumeration:
+```rust
+enum Monotonicity {
+    None,
+    Physical,
+    Logical
+}
+```
+
+1. When creating a plan for `SELECT` queries, we upgrade the monotonicity information of sources to *logical monotonicity*, unless they're physically monotonic, in which case they're not changed.
+2. The optimizer uses this information to derive the monotonicity information for plans, and ensures that all operators can expect data fulfilling their monotonicity requirements.
+3. Rendering removes defensive consolidate operators.
+
+# Rollout
+[rollout]: #rollout
+
+<!--
+Describe what steps are necessary to enable this feature for users.
+How do we validate that the feature performs as expected? What monitoring and observability does it require?
+-->
+
+To enable this feature in production we should take the following steps:
+1. Implement the feature, but guard it with a flag.
+2. Adjust tests based on what's discussed above.
+3. Enable the feature in staging environments.
+4. Enable the feature in production environments.
+5. Remove the feature flag.
+
+## Testing and observability
+[testing-and-observability]: #testing-and-observability
+
+<!--
+Testability and explainability are top-tier design concerns!
+Describe how you will test and roll out the implementation.
+When the deliverable is a refactoring, the existing tests may be sufficient.
+When the deliverable is a new feature, new tests are imperative.
+
+Describe what metrics can be used to monitor and observe the feature.
+What information do we need to expose internally, and what information is interesting to the user?
+How do we expose the information?
+
+Basic guidelines:
+
+* Nearly every feature requires either Rust unit tests, sqllogictest tests, or testdrive tests.
+* Features that interact with Kubernetes additionally need a cloudtest test.
+* Features that interact with external systems additionally should be tested manually in a staging environment.
+* Features or changes to performance-critical parts of the system should be load tested.
+-->
+
+## Lifecycle
+[lifecycle]: #lifecycle
+
+<!--
+If the design is risky or has the potential to be destabilizing, you should plan to roll the implementation out behind a feature flag.
+List all feature flags, their behavior and when it is safe to change their value.
+Describe the [lifecycle of the feature](https://www.notion.so/Feature-lifecycle-2fb13301803b4b7e9ba0868238bd4cfb).
+Will it start as an alpha feature behind a feature flag?
+What level of testing will be required to promote to beta?
+To stable?
+-->
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+<!--
+Why should we *not* do this?
+-->
+
+The feature adds complexity to the optimizer and tightens the contract between the optimizer and rendering.  This imposes some risk to surface undefined behavior, which could in the worst case lead to crashes or incorrect data.
```

Comment:
As we discussed in the timely office hours today, for me the main risk is that the new monotonic analysis (which happens during the MIR stage of the pipeline) will encode information that depends on the properties of the physical (LIR) plans derived during `mir_to_lir` lowering and more specifically aspects of how they are rendered.

This makes everything quite brittle, as over time various pieces might change:

1. The monotonic analysis can be out of sync with the monotonicity properties advertised by the different LIR operators.
2. The monotonicity properties advertised by LIR operators might be out of sync with their rendering logic.

If things go out of sync because of (1) and (2), we might start producing incorrect answers without noticing. Therefore, it will be great if we can design the code with such changes in mind.

I think it might be good idea to ask @mgree about PL techniques for achieving consistency between the various IR layers.

Reply:
I was talking about this a bit on Friday with @antiguru. I don't know of any lightweight techniques that give strong guarantees.

Research compilers written in dependently typed languages can verify these properties. That's... a lot.

It's possible to use phantom types to get some static guarantees (i.e., have a type index on MIR or LIR terms that describe each expression's properties, with constraints on how they're composed). The phantom types don't _do_ anything, so it'd be up to us to build 'safe' ways of constructing and handling them. You get a single audit point---the safe operations---but it makes it much more annoying to work with the structures.

A third option is to try to make all of this logic/reasoning explicit at the MIR level, e.g., with AST nodes that explicitly mark inputs as physically monotonic or whole queries as one-shot. You won't get static guarantees, but you can dynamically check that the code that lowers to LIR is getting a consistent view. I suspect that some small annotations would be enough to build a type system that tracked more than column types---like indices/arrangements or monotonicity.

Reply:
Given the fact that we have avoided a type-safe IR until now I think that option (3) probably would be the least controversial.

> e.g., with AST nodes that explicitly mark inputs as physically monotonic or whole queries as one-shot

Do you mean adding variants such as `MirRelationExpr::OneShot` or a field attached to all variants (`one_shot: bool`).

Reply:
I imagined... both. I'd add `MirRelationExpr` a variant indicating that parts of the term are one-shot; I'd also add a field to `Get` to indicate physically monotonic data. Neither approach is perfect: just marking `Get`s, you'd need to recompute how far out the use of monotonic data stays monotonic; marking expressions means optimizer passes might have to remember whether or not they were operating in a `MRE::OneShot` context.